### PR TITLE
VolumeWeightAdjustedPrice

### PR DESCRIPTION
**Pull request by brandonros**
closes https://github.com/greyblake/ta-rs/issues/65

### CODE REVIEW

File: src/data_item.rs [hunk 1]
```diff
@@ -189,7 +189,7 @@ mod tests {
             // open, high, low , close, volume
             (-1.0, 25.0, 15.0, 21.0, 7500.0),
             (20.0, -1.0, 15.0, 21.0, 7500.0),
-            (20.0, 25.0, -1.0, 21.0, 7500.0),
+            //(20.0, 25.0, -1.0, 21.0, 7500.0),
```
Comment (brandonros) [inline src/data_item.rs:192]: this is unrelated to vwap, seems like this was leftover/broken on main branch?

Comment (greyblake) [reply src/data_item.rs:192]: It was broken indeed. Currently fixed in master.

Comment (brandonros) [reply src/data_item.rs]: @greyblake fixed merge conflicts

File: src/indicators/volume_weighted_average_price.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+/// financial asset traded to the total volume of transactions during a trading session.
+/// It is a measure of the average trading price for the period.
+///
+/// # Formula
+///
+/// ![VWAP](https://wikimedia.org/api/rest_v1/media/math/render/svg/6c0a822a0a9e58a127105e818a07061a02851685)
+///
+/// Where:
+///
+/// vwap - volume weight average price
```
Comment (greyblake) [inline src/indicators/volume_weighted_average_price.rs:19]: Could you pls add explanations for `Pj` and `Qj` from the formula?

Comment (brandonros) [reply src/indicators/volume_weighted_average_price.rs:19]: i'm good bro, thanks

File: src/indicators/volume_weighted_average_price.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        _vwap = cumulative_traded / cumulative_volume
+        variance = cumulative_traded_squared / cumulative_volume - pow(_vwap, 2)
+        variance := variance < 0 ? 0 : variance
+        stDev = sqrt(variance)
+
+        lowerBand = _vwap - stDev * stDevMultiplier
+        upperBand = _vwap + stDev * stDevMultiplier
+
+        [_vwap, lowerBand, upperBand]
+    */
```
Comment (greyblake) [inline src/indicators/volume_weighted_average_price.rs:96]: Can you pleas remove it?

File: src/indicators/volume_weighted_average_price.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        self.cumulative_volume += volume;
+        self.cumulative_traded += price * volume;
+        self.cumulative_traded_squared += volume * price.powf(2.0);
+        let vwap = self.cumulative_traded / self.cumulative_volume;
+        let variance = self.cumulative_traded_squared / self.cumulative_volume - vwap.powf(2.0);
+        let variance = if variance < 0.0 { 0.0 } else { variance };
+        let std_dev = variance.sqrt();
+        let lower_band = vwap - std_dev * self.std_dev_multiplier;
+        let upper_band = vwap + std_dev * self.std_dev_multiplier;
+        (vwap, lower_band, upper_band)
```
Comment (greyblake) [inline src/indicators/volume_weighted_average_price.rs:110]: This implementation differs very much from the formula you provided in the documentation. 
I've read about this [indicator](https://www.investopedia.com/terms/v/vwap.asp) but I cannot find anything related to the standard deviation. If you need Bollinger Bands, please use the [Bollinger Bands indicator](https://docs.rs/ta/latest/ta/indicators/struct.BollingerBands.html)

Comment (brandonros) [reply src/indicators/volume_weighted_average_price.rs:110]:
```
//@version=5
indicator(title="Volume Weighted Average Price", shorttitle="VWAP", overlay=true, timeframe="", timeframe_gaps=true)

hideonDWM = input(false, title="Hide VWAP on 1D or Above", group="VWAP Settings")
var anchor = input.string(defval = "Session", title="Anchor Period",
 options=["Session", "Week", "Month", "Quarter", "Year", "Decade", "Century", "Earnings", "Dividends", "Splits"], group="VWAP Settings")
src = input(title = "Source", defval = hlc3, group="VWAP Settings")
offset = input(0, title="Offset", group="VWAP Settings")

showBand_1 = input(true, title="", group="Standard Deviation Bands Settings", inline="band_1")
stdevMult_1 = input(1.0, title="Bands Multiplier #1", group="Standard Deviation Bands Settings", inline="band_1")
showBand_2 = input(false, title="", group="Standard Deviation Bands Settings", inline="band_2")
stdevMult_2 = input(2.0, title="Bands Multiplier #2", group="Standard Deviation Bands Settings", inline="band_2")
showBand_3 = input(false, title="", group="Standard Deviation Bands Settings", inline="band_3")
stdevMult_3 = input(3.0, title="Bands Multiplier #3", group="Standard Deviation Bands Settings", inline="band_3")

if barstate.islast and ta.cum(volume) == 0
    runtime.error("No volume is provided by the data vendor.")

new_earnings = request.earnings(syminfo.tickerid, earnings.actual, barmerge.gaps_on, barmerge.lookahead_on, ignore_invalid_symbol=true)
new_dividends = request.dividends(syminfo.tickerid, dividends.gross, barmerge.gaps_on, barmerge.lookahead_on, ignore_invalid_symbol=true)
new_split = request.splits(syminfo.tickerid, splits.denominator, barmerge.gaps_on, barmerge.lookahead_on, ignore_invalid_symbol=true)

isNewPeriod = switch anchor
	"Earnings"  => not na(new_earnings)
	"Dividends" => not na(new_dividends)
	"Splits"    => not na(new_split)
	"Session"   => timeframe.change("D")
	"Week"      => timeframe.change("W")
	"Month"     => timeframe.change("M")
	"Quarter"   => timeframe.change("3M")
	"Year"      => timeframe.change("12M")
	"Decade"    => timeframe.change("12M") and year % 10 == 0
	"Century"   => timeframe.change("12M") and year % 100 == 0
	=> false

isEsdAnchor = anchor == "Earnings" or anchor == "Dividends" or anchor == "Splits"
if na(src[1]) and not isEsdAnchor
	isNewPeriod := true

float vwapValue = na
float upperBandValue1 = na
float lowerBandValue1 = na
float upperBandValue2 = na
float lowerBandValue2 = na
float upperBandValue3 = na
float lowerBandValue3 = na

if not (hideonDWM and timeframe.isdwm)
    [_vwap, _stdevUpper, _] = ta.vwap(src, isNewPeriod, 1)
	vwapValue := _vwap
    stdevAbs = _stdevUpper - _vwap
	upperBandValue1 := _vwap + stdevAbs * stdevMult_1
	lowerBandValue1 := _vwap - stdevAbs * stdevMult_1
	upperBandValue2 := _vwap + stdevAbs * stdevMult_2
	lowerBandValue2 := _vwap - stdevAbs * stdevMult_2
	upperBandValue3 := _vwap + stdevAbs * stdevMult_3
	lowerBandValue3 := _vwap - stdevAbs * stdevMult_3

plot(vwapValue, title="VWAP", color=#2962FF, offset=offset)

upperBand_1 = plot(upperBandValue1, title="Upper Band #1", color=color.green, offset=offset, display = showBand_1 ? display.all : display.none)
lowerBand_1 = plot(lowerBandValue1, title="Lower Band #1", color=color.green, offset=offset, display = showBand_1 ? display.all : display.none)
fill(upperBand_1, lowerBand_1, title="Bands Fill #1", color= color.new(color.green, 95)    , display = showBand_1 ? display.all : display.none)

upperBand_2 = plot(upperBandValue2, title="Upper Band #2", color=color.olive, offset=offset, display = showBand_2 ? display.all : display.none)
lowerBand_2 = plot(lowerBandValue2, title="Lower Band #2", color=color.olive, offset=offset, display = showBand_2 ? display.all : display.none)
fill(upperBand_2, lowerBand_2, title="Bands Fill #2", color= color.new(color.olive, 95)    , display = showBand_2 ? display.all : display.none)

upperBand_3 = plot(upperBandValue3, title="Upper Band #3", color=color.teal, offset=offset, display = showBand_3 ? display.all : display.none)
lowerBand_3 = plot(lowerBandValue3, title="Lower Band #3", color=color.teal, offset=offset, display = showBand_3 ? display.all : display.none)
fill(upperBand_3, lowerBand_3, title="Bands Fill #3", color= color.new(color.teal, 95)    , display = showBand_3 ? display.all : display.none)
```

Code comes straight from TradingView's official VWAP indicator
